src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -61,20 +64,35 @@ public void run() {
    */
   private void runNodes(List<T> freeNodes) {
     List<IWorker<T>> runnables = m_factory.createWorkers(freeNodes);
+    CountDownLatch latch = null;
+    if (useLatches && runnables.size() > 1) {
+      latch = new CountDownLatch(runnables.size());
+    }
+    boolean latchUsed = false;
     for (IWorker<T> r : runnables) {
       m_activeRunnables.add(r);
+      if (latch != null && (r instanceof TestMethodWorker)) {
+        //Being conservative here and setting this only for test method workers
+        r.setLatch(latch);
+        latchUsed = true;
+      }
       ppp("Added to active runnable");
       setStatus(r, Status.RUNNING);
       ppp("Executing: " + r);
       try {
         execute(r);
-//        if (m_threadCount > 1) execute(r);
-//        else r.run();
       }
       catch(Exception ex) {
         ex.printStackTrace();
       }
     }
+    if (latch != null && latchUsed) {
+      try {
+        latch.await(10, TimeUnit.SECONDS);
```

Comment:
@juherr - Fair enough. I have now changed the logic and did the following 
* Extracted the latch adding behavior into its own interface that only `TestMethodWorker` implements. Now `GraphThreadPoolExecutor` runs an instance check and if it finds its an instance that can accept the latch, it goes on to set the latch to the worker. Please see if this would sound convincing.